### ISSUE TITLE
Link to code page on individual course overview

### ISSUE THREAD

**Issue by SebastianZug** [liaScript/LiaScript#11]
Hi everyone,

The course overview (reachable by the home button) is essential to receive the basic information on all LiaScript courses. Beside delete and reset buttons, a link on the code of a course would be helpful. 

When I asked my students for individual extensions of the material, they explained to me that they were not able to find the course on GitHub. A direct link would overcome this excuse :-)

Sebastian

**Comment by andre-dietrich** [comment]
Hello Sebastian,

the basic idea was, to enable anyone to host her/his course everywhere, but of course due to some cors limitations, the list is currently limited to GitHub and Dropbox. It would be great to add more solutions to this list or we simply need to use a server as a proxy. 

https://github.com/liaScript/LiaScript/wiki/Hosting-List

But at the moment I would like to figure out, how far we can get without servers or a centralized infrastructure (serverless LMS) ;)

**Comment by pwab** [comment]
> I mean, this is the reason, why there is currently no link to the GitHub project itself, since other ways for storing content could also be utilized... at least, I hope so ..

Wouldn't it be possible to just detect known providers and convert the raw link back to the repository link? As long as it is known how GitHub and others provide raw-links.
So for Github contet this would mean to change the domain to github.com and drop the filename and branch from the string. This would convert:

`https://liascript.github.io/course/?https://raw.githubusercontent.com/pwab/LiaTests/main/README.md`

to

`https://github.com/pwab/LiaTests/`

Another option would be to add a liascript command to the head where you could provide a link to the sources which then is shown in the course overview. This would allow pasting links which lead to higher level webpages (like course collections) and no conversion logic would be needed.

**Comment by andre-dietrich** [comment]
Thanks, this is actually a great idea, since it works atm the other way around, you can refer to the branch/README.md and it gets internally translated into the raw format. (The change of the default master to main branch had made it a bit more complicated). I have to check how this might affect the internal state of the application, but it is definitely an improvement and shrinks to looooong url...

**Comment by BeastyBlacksmith** [comment]
> but of course due to some cors limitations,

I think I am running into this trying to host my course on a institution hosted gitlab instance (you can view the course [here](https://comp-bio.idmpages.uni-h.de/teaching/julia-physics-report/?README.md#7)), can you say a bit more about this?. When running the dev-server locally everything works fine, but on the server it doesn't fetches the files and can't find the figures (which is related to path issues, I suspect).

**Comment by andre-dietrich** [comment]
I checked the link and I could not find a Markdown-resource yet. The resulting page is delivered as an rendered HTML, although the URL does look different.

This for example is an older project from gitlab:

https://gitlab.com/OvGU-ESS/SelectScript/-/blob/master/README.md

However, this URL cannot be loaded directly, since it does not point to the Markdown file directly. If I want to refer to the Markdown-file, I do have to use the following URL. If you click on it, you will get the Markdown-file as plain text. This is what LiaScript requires.

https://gitlab.com/OvGU-ESS/SelectScript/-/raw/master/README.md

However number 2, this URL cannot be loaded directly via:

https://liascript.github.io/course/?https://gitlab.com/OvGU-ESS/SelectScript/-/raw/master/README.md

This will fail in the first trial, with the following error message:

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://gitlab.com/OvGU-ESS/SelectScript/-/raw/master/README.md. (Reason: CORS request did not succeed). Status code: (null).

Although the file is delivered, the HTTP response header will have a value that does not allow the received content to be loaded from another URL, in this case the liascript-page:

https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

This requires to change the server settings...

But, since not everyone is allowed to make this changes, LiaScript makes a second attempt and tries to load the content via a CORS proxy. Thus the final URL will look like this:

https://liascript.github.io/course/?https://api.allorigins.win/raw?url=https://gitlab.com/OvGU-ESS/SelectScript/-/raw/master/README.md#1

If you do not like this one, you can manually also add another proxy ...

If you are running the dev-server from local host, the browser is more willing to ignore CORS related issues because it expects that you know what you are doing. Doing the same online via HTTPS might be a risk, which is why the browser will be blocking more stuff...

**Comment by andre-dietrich** [comment]
@BeastyBlacksmith could you share also the URL to your project, so that we could test it...

**Comment by BeastyBlacksmith** [comment]
Ah, sorry I tried something else, thus a different page, Markdown file is this one: https://gitlab.uni-hannover.de/comp-bio/teaching/julia-physics-report/-/blob/master/README.md

**Comment by andre-dietrich** [comment]
Like in GitHub, there is an open-raw button, which seems to provide the right URL.

https://liascript.github.io/course/?https://gitlab.uni-hannover.de/comp-bio/teaching/julia-physics-report/-/raw/master/README.md

However, this will not work on the first trial and LiaScript will make automatically a second attempt with the following URL ... Which succeeded in my case ;-)

https://liascript.github.io/course/?https://api.allorigins.win/raw?url=https://gitlab.uni-hannover.de/comp-bio/teaching/julia-physics-report/-/raw/master/README.md

**Comment by BeastyBlacksmith** [comment]
Yes, thats working (even the figures and pdf-files), but fetching of the code files does not work ( e.g. when expanding the `data/tau-M.jl` box on [slide 7](https://liascript.github.io/course/?https://api.allorigins.win/raw?url=https://gitlab.uni-hannover.de/comp-bio/teaching/julia-physics-report/-/raw/master/README.md#7))

**Comment by andre-dietrich** [comment]
In this case LiaScript is not aware that this is a location, it operates only with strings. Expanding this, would be an interesting feature ;-)
However, you can use a workaround in your script...

```js
let file = window.location.search.slice(1).replace("README.md", "working_directory/") + "data/tau-M.jl"

// or

let file = window.location.search.slice(1).replace("README.md", "working_directory/") + "@0"

// ------------------------------------------------------------
// => https://api.allorigins.win/raw?url=https://gitlab.uni-hannover.de/comp-bio/teaching/julia-physics-report/-/raw/master/working_directory/data/tau-M.jl
``` 

This will create a URL based on the parameter that loads your course. When you move your course to another location, this will still work ...

**Comment by BeastyBlacksmith** [comment]
I can get this working on the devserver, but it it doesn't even print the location to the console if I run it on the liascript server..

**Comment by andre-dietrich** [comment]
Probably you could try an `alert` or `window.console.log` to test the output?
